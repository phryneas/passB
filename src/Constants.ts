export enum executionContext {
  background = 'background',
  popup = 'popup',
  options = 'options',
}

export const latestHostAppVersion = '1.0.1';
