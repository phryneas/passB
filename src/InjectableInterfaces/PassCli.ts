export interface PassCli {
  list(): Promise<string[]>;
  show(path: string): Promise<string[]>;
  getHostAppVersion(): Promise<string>;
}
