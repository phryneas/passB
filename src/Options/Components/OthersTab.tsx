import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  List,
  ListItem,
  ListItemIcon,
  ListItemText,
  Typography,
} from 'material-ui';
import {Cancel, Check, Clear} from 'material-ui-icons';
import * as React from 'react';
import {Interfaces, Symbols} from 'Container';
import {lazyInject} from 'Decorators/lazyInject';

interface State {
  showResetDialog: boolean;
}

export class OthersTab extends React.Component<{}, State> {
  public state: State = {
    showResetDialog: false,
  };

  @lazyInject(Symbols.PassB)
  private passB: Interfaces.PassB;

  public render(): JSX.Element {
    const {
      showResetDialog,
    } = this.state;

    return (
      <div>
        <List>
          <ListItem
            button={true}
            onClick={() => this.setState({showResetDialog: true})}
          >
            <ListItemIcon><Clear/></ListItemIcon>
            <ListItemText primary={browser.i18n.getMessage('options_reset_extension')}/>
          </ListItem>
        </List>
        <Dialog
          onClose={() => this.setState({showResetDialog: false})}
          maxWidth="xs"
          open={showResetDialog}
        >
          <DialogContent>
            <Typography type="subheading">{browser.i18n.getMessage('options_reset_extension_confirm')}</Typography>
          </DialogContent>
          <DialogActions>
            <Button onClick={() => this.setState({showResetDialog: false})} color="primary">
              <Cancel/>
            </Button>
            <Button
              onClick={() => {
                browser.storage.local.clear();
                this.passB.reloadExtension();
                browser.tabs.getCurrent().then((tab: browser.tabs.Tab) => {
                  if (tab && tab.id) {
                    browser.tabs.remove(tab.id);
                  }
                });
              }}
              color="primary"
            >
              <Check/>
            </Button>
          </DialogActions>
        </Dialog>
      </div>
    );
  }
}
