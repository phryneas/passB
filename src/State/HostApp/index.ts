import {createSelector} from 'reselect';
import {actionCreatorFactory} from 'typescript-fsa';
import {reducerWithInitialState} from 'typescript-fsa-reducers';
import {StoreContents} from 'InjectableInterfaces/State';
import {getHostApState} from 'State/Selectors';

// interfaces
export interface Error {
  message: string;
  type: ErrorType;
}

export interface HostAppState {
  version: string;
  lastError?: Error;
}

export enum ErrorType {
  PASS_EXECUTION_ERROR = 'PASS_EXECUTION_ERROR',
  HOST_APP_ERROR = 'HOST_APP_ERROR',
}

// actions
const actionCreator = actionCreatorFactory('HOST_APP');
export const setVersion = actionCreator<string>('SET_VERSION');
export const setLastError = actionCreator<Error>('SET_LAST_ERROR');
export const clearLastError = actionCreator<void>('CLEAR_LAST_ERROR');

export const Actions = {
  setVersion,
  setLastError,
  clearLastError,
};

// reducer
export const initialState = {
  version: 'unknown',
};
export const reducer = reducerWithInitialState(initialState)
  .case(setVersion, (state: HostAppState, version: string): HostAppState => ({...state, version}))
  .case(setLastError, (state: HostAppState, lastError: Error): HostAppState => ({...state, lastError}))
  .case(clearLastError, ({lastError, ...stateRest}: HostAppState): HostAppState => stateRest)
  .build();

// selectors
export const getLastError: (state: StoreContents) => Error | undefined =
  createSelector(getHostApState, (state: HostAppState) => state.lastError);

export const getVersion: (state: StoreContents) => string =
  createSelector(getHostApState, (state: HostAppState) => state.version);
