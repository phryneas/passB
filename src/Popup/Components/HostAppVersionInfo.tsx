import {Button, List, ListItem, ListItemText, Popover} from 'material-ui';
import {Warning} from 'material-ui-icons';
import * as React from 'react';
import {connect} from 'react-redux';
import {lt, valid} from 'semver';
import {latestHostAppVersion} from 'Constants';
import {StoreContents} from 'InjectableInterfaces/State';
import {getVersion} from 'State/HostApp/index';

export interface Props {
  visible: boolean;
  hostAppVersion: string;
}

export interface State {
  open: boolean;
}

export class HostAppVersionInfoComponent extends React.Component<Props, State> {
  public state: State = {
    open: false,
  };

  private button: HTMLButtonElement;

  public render(): JSX.Element[] | false {
    const {visible, hostAppVersion} = this.props;
    const {open} = this.state;

    if (!visible) {
      return false;
    }

    return [(
      <Button
        rootRef={(button: HTMLButtonElement) => this.button = button}
        key="btn"
        color="accent"
        onClick={() => this.setState({open: true})}
      >
        <Warning/>
      </Button>
    ),      (
      <Popover
        key="popover"
        anchorEl={this.button}
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'center',
        }}
        transformOrigin={{
          vertical: 'top',
          horizontal: 'center',
        }}
        open={open}
        onClose={() => this.setState({open: false})}
      >
            <List>
              <ListItem>
                <ListItemText primary={browser.i18n.getMessage('popup_host_app_nag_update', [hostAppVersion, latestHostAppVersion])} />
              </ListItem>
              <ListItem
                button={true}
                className="hostAppInfo"
                onClick={() => browser.tabs.create({url: 'https://passb.github.io/host_application.html'})}
              >
                <ListItemText primary={browser.i18n.getMessage('popup_host_app_error_option_install')}/>
              </ListItem>
            </List>
      </Popover>
    )];
  }
}

export const mapStateToProps = (state: StoreContents): Props => {
  const version = getVersion(state);
  return {
    hostAppVersion: version,
    visible: valid(version) !== null && lt(version, latestHostAppVersion),
  };
};

export const HostAppVersionInfo = connect(mapStateToProps)(HostAppVersionInfoComponent);
