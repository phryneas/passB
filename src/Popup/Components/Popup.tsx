import {AppBar, Button, Card, CardContent, Toolbar, Typography, WithStyles} from 'material-ui';
import {Refresh, Settings} from 'material-ui-icons';
import {withStyles, StyleRules} from 'material-ui/styles';
import * as React from 'react';
import Tab = browser.tabs.Tab;
import {MemoryRouter, Route, RouteComponentProps, RouteProps, Switch} from 'react-router';
import {Interfaces, Symbols} from 'Container';
import {lazyInject} from 'Decorators/lazyInject';
import {ErrorBoundary} from './ErrorBoundary';
import {HostAppErrorWrapper} from './HostAppErrorWrapper';
import {HostAppVersionInfo} from './HostAppVersionInfo';
import {ListView} from './Views/ListView';

interface Props {
}

interface State {
  activeTab?: Tab;
}

const styles: StyleRules = {
  flexLeft: {
    flex: 1,
    marginLeft: '15px',
  },
  cardSize: {
    maxHeight: '390px',
    overflow: 'auto',
  },
};

class ClassLessPopup extends React.Component<Props & WithStyles<keyof typeof styles>, State> {
  public state: State = {};
  private gatheredRoutes: RouteProps[] = this.gatherRoutes();

  @lazyInject(Symbols.PassB)
  private passB: Interfaces.PassB;

  public componentDidMount(): void {
    browser.tabs.query({
      active: true,
      currentWindow: true,
    }).then((tabs: Tab[]) => this.setState({activeTab: tabs[0]}));
  }

  public render(): JSX.Element {
    const {activeTab} = this.state;
    const {classes} = this.props;

    return (
      <MemoryRouter>
        <div>
          <AppBar position="fixed">
            <Toolbar>
              <Typography type="title" color="inherit" className={classes.flexLeft}>
                {browser.i18n.getMessage('extensionName')}
              </Typography>
              <HostAppVersionInfo />
              <Button
                color="contrast"
                onClick={() => {
                  this.passB.reloadEntries();
                  window.close();
                }}
              >
                <Refresh/>
              </Button>
              <Button
                color="contrast"
                onClick={() => {
                  browser.runtime.openOptionsPage();
                  window.close();
                }}
              >
                <Settings/>
              </Button>
            </Toolbar>
          </AppBar>
          <Card>
            <CardContent className={classes.cardSize}>
              <ErrorBoundary>
                <HostAppErrorWrapper>
                  <Switch>
                    {this.gatheredRoutes.map((route: RouteProps) => <Route key={String(route.path)}  {...route} />)}
                    <Route
                      render={
                        ({history}: RouteComponentProps<{}>) => (
                          <ListView
                            url={activeTab && activeTab.url ? activeTab.url : ''}
                          />
                        )}
                    />
                  </Switch>
                </HostAppErrorWrapper>
              </ErrorBoundary>
            </CardContent>
          </Card>
        </div>
      </MemoryRouter>
    );
  }

  private gatherRoutes(): RouteProps[] {
    const gatheredRoutes: RouteProps[] = [];
    for (const extension of this.passB.getAllExtensions()) {
      for (const route of extension.routes) {
        if (!(route.path || '').startsWith(`/extension/${extension.name}/`)) {
          throw Error(
            `every route path for extension ${extension.name} has to start with "/extension/${extension.name}/"`,
          );
        }
        gatheredRoutes.push(route);
      }
    }

    return gatheredRoutes;
  }
}

export const Popup = withStyles<keyof typeof styles>(styles)(ClassLessPopup);
