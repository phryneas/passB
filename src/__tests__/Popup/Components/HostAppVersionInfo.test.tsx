import {shallow} from 'enzyme';
import * as React from 'react';

jest.mock('Constants', () => ({
  latestHostAppVersion: '1.5.1',
}));

import {mapStateToProps, HostAppVersionInfoComponent} from 'Popup/Components/HostAppVersionInfo';
import {reducer, Actions} from 'State/HostApp';

import {} from 'Constants';
import {StoreContents} from 'InjectableInterfaces/State';

describe('HostAppVersionInfo', () => {
  describe('HostAppVersionInfoComponent', () => {
    it('renders empty on visible=false', () => {
      expect(shallow(
        <HostAppVersionInfoComponent visible={false} hostAppVersion={'unknown'}/>,
      )).toMatchSnapshot();
    });

    describe('visible', () => {
      it('renders', () => {
        expect(shallow(
          <HostAppVersionInfoComponent visible={true} hostAppVersion={'1.4.9'}/>,
        )).toMatchSnapshot();
      });
    });
  });
});

describe('mapStateToProps', () => {
  it('installed version: unknown', () => {
    const state = {hostApp: reducer(undefined!, {type: 'other'})};
    expect(mapStateToProps(state as StoreContents)).toMatchSnapshot();
  });

  it('installed version: lower', () => {
    const state = {hostApp: reducer(undefined!, Actions.setVersion('1.5.0'))};
    expect(mapStateToProps(state as StoreContents)).toMatchSnapshot();
  });

  it('installed version: current', () => {
    const state = {hostApp: reducer(undefined!, Actions.setVersion('1.5.1'))};
    expect(mapStateToProps(state as StoreContents)).toMatchSnapshot();
  });

  it('installed version: higher', () => {
    const state = {hostApp: reducer(undefined!, Actions.setVersion('1.5.2'))};
    expect(mapStateToProps(state as StoreContents)).toMatchSnapshot();
  });
});
