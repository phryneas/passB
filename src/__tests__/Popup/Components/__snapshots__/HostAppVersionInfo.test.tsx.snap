// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HostAppVersionInfo HostAppVersionInfoComponent renders empty on visible=false 1`] = `""`;

exports[`HostAppVersionInfo HostAppVersionInfoComponent visible renders 1`] = `
<div>
  <withStyles(Button)
    color="accent"
    key="btn"
    onClick={[Function]}
    rootRef={[Function]}
  >
    <pure(Warning) />
  </withStyles(Button)>
  <withStyles(Popover)
    anchorOrigin={
      Object {
        "horizontal": "center",
        "vertical": "bottom",
      }
    }
    key="popover"
    onClose={[Function]}
    open={false}
    transformOrigin={
      Object {
        "horizontal": "center",
        "vertical": "top",
      }
    }
  >
    <withStyles(List)>
      <withStyles(ListItem)>
        <withStyles(ListItemText)
          primary="Your installed hostApp version is $currentVersion$ while the latest version is $latestVersion$. Some things might not work correctly until you upgrade."
        />
      </withStyles(ListItem)>
      <withStyles(ListItem)
        button={true}
        className="hostAppInfo"
        onClick={[Function]}
      >
        <withStyles(ListItemText)
          primary="install the host application"
        />
      </withStyles(ListItem)>
    </withStyles(List)>
  </withStyles(Popover)>
</div>
`;

exports[`mapStateToProps installed version: current 1`] = `
Object {
  "hostAppVersion": "1.5.1",
  "visible": false,
}
`;

exports[`mapStateToProps installed version: higher 1`] = `
Object {
  "hostAppVersion": "1.5.2",
  "visible": false,
}
`;

exports[`mapStateToProps installed version: lower 1`] = `
Object {
  "hostAppVersion": "1.5.0",
  "visible": true,
}
`;

exports[`mapStateToProps installed version: unknown 1`] = `
Object {
  "hostAppVersion": "unknown",
  "visible": false,
}
`;
